[TOOLS-4704] when migration using SQL, the target schema enters null

Kang-dot opened the pull request:
http://jira.cubrid.org/browse/TOOLS-4704

- **Purpose**
SQL을 이용한 이관 기능 사용 시 schema 정보가 누락되어 쿼리가 정확하게 수행이 안되는 문제를 수정합니다

- **Implementation**
SQL을 생성하는데 사용하는 object에 owner 정보를 포함할 수 있도록 class를 수정합니다
추가로 owner를 선택할 수 있도록 combo box를 추가합니다

- **Remarks**
csv source 이관 기능이 현재 비활성 상태입니다. 이번 이슈를 수정하면서 관련된 기능이 영향을 받았을 확률이 있습니다. 추후 기능을 활성화 시킬 경우 문제가 될 수 있으니 참고해 주세요

Kang-dot committed:
[RND-2276] add schema select combo box

Kang-dot committed:
[TOOLS-4704] when migration using SQL, the target schema enters null

- user schema is not applied to sql extraction part, correcting

Kang-dot committed:
[TOOLS-4704] adding and delete comments

- remote unused code line
- add description in method

Kang-dot committed:
[TOOLS-4704] Modified to work in script environment

Kang-dot committed:
[TOOLS-4704] Modifying the combobox to work correctly

- Modified to work correctly for GUI script execution

Kang-dot committed:
[TOOLS-4704] improvement code

Kang-dot committed:
[TOOLS-4704] formatting flies

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/view/SQLTableManageView.java:
```diff
@@ -588,13 +612,40 @@ public void run(IProgressMonitor monitor)
      */
     public VerifyResultMessages save() {
         // Check
+
```
의미가 있는 줄 띄움인가요?

Kang-dot replied:
개인 습관으로 인해 의미 없는 줄 바꿈이 발생한 것 같습니다. 삭제하겠습니다.

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/view/SQLTableManageView.java:
```diff
@@ -310,27 +317,39 @@ protected void createControl(Composite parent) {
                 new CellEditorFactory[] {
                     new TextCellEditorFactory(),
                     null,
+                    //                    null,
```
의미 없는 주석으로 보입니다.

Kang-dot replied:
주석 삭제 과정에서 누락된 것으로 보입니다. 수정하겠습니다.

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/view/SQLTableManageView.java:
```diff
@@ -588,13 +612,40 @@ public void run(IProgressMonitor monitor)
 
+    /** setup real combo box value(target schema name list) */
+    protected void setupCombobox() {
+
+        String[] schemaNameArr =
+                config.getTarCatalog().getSchemas().stream()
+                        .map(Schema::getName)
+                        .toArray(String[]::new);
+
+        this.tarSchemaList = schemaNameArr;
+
+        CellEditor[] cellEditorArray = tvSQL.getCellEditors();
+
+        ((ComboBoxCellEditor) cellEditorArray[2]).setItems(schemaNameArr);
```
NIT: 코드 포맷터가 자동으로 한 줄씩 띄운 것인가요? 그게 아니라면 각 코드를 한 줄 띄운 이유가 있는지 궁금합니다.

Kang-dot replied:
618 ~ 621줄은 formatter 작동 결과이나 623 ~ 627줄은 개인 습관에 의한 개행입니다. 삭제하겠습니다

Srltas commented on com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/core/engine/config/MigrationConfiguration.java:
```diff
@@ -4947,6 +4965,22 @@ public void setSrcCatalog(Catalog srcCatalog, boolean reset) {
+     */
+    public void setTarCatalog(Catalog tarCatalog) {
+        if (tarCatalog == null) {
+            throw new IllegalArgumentException("Catalog can't not be null.");
```
```suggestion
            throw new IllegalArgumentException("Catalog cannot be null.");
```

Kang-dot replied:
오탈자 수정해서 반영하겠습니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/ObjectMappingPage.java:
```diff
@@ -191,6 +191,7 @@ protected void afterShowCurrentPage(PageChangedEvent event) {
 
             showLobInfo(sourceCatalog);
             cfg.setSrcCatalog(sourceCatalog, isFirstVisible && !mw.isLoadMigrationScript());
+            cfg.setTarCatalog(mw.getTargetCatalog());
```
`getTargetCatalog()`는 반환값이 null 일 수 있습니다. null에 대한 처리가 필요해 보입니다.

Kang-dot replied:
target catalog가 null인 경우는 target이 file 출력일 때의 경우인 것으로 알고 있습니다 추후 file 출력 관련 수정할 때 반영 해도 될 것 같습니다만 우선 추가로 분석해 보겠습니다.

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -774,8 +811,12 @@ protected void checkTableIsInTargetDb(
         if (config.targetIsOnline() && !wizardStatus.isTargetOfflineMode()) {
-            final boolean contained =
-                    existsTableNameList.contains(targetTable.getName().toUpperCase(Locale.US));
+            boolean contained =
+                    Optional.<List<String>>ofNullable(
+                                    existsTableNameList.get(setc.getTargetOwner()))
+                            .orElse(Collections.emptyList()).stream()
+                            .anyMatch(targetTable.getName()::equals);
+
```
단순한 null 체크 및 포함 여부 확인을 위해 Optional과 Stream을 사용하는데, 오버 엔지니어링으로 보입니다.
Map과 List에 있는 메소드를 이용해 간단히 구현하는 것은 어떤가요?

```java
boolean contained = existsTableNameList
        .getOrDefault(setc.getTargetOwner(), Collections.emptyList())
        .contains(targetTable.getName());
```

Kang-dot replied:
제시해주신 코드로 수정하겠습니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -209,7 +210,7 @@ public static boolean verifyTargetDBObjName(String name) {
 
     private final List<String> existsViewNameList = new ArrayList<String>();
-    private final List<String> existsTableNameList = new ArrayList<String>();
+    private final HashMap<String, List<String>> existsTableNameList = new HashMap<>();
```
변수 타입은 HashMap 보단 Map을 사용하는 것이 좋겠습니다.

Kang-dot replied:
Map으로 수정하겠습니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -272,39 +279,69 @@ protected VerifyResultMessages checkCSVCfg(MigrationConfiguration config) {
+                            .add(scc.getTarget());
+                    csvOnlyCreateTarget
+                            .computeIfAbsent(scc.getTargetOwner(), val -> new ArrayList<String>())
+                            .add(scc.getTarget());
```
수정 전 코드에서는 `csvOnlyCreateTarget.remove()` 메소드를 사용했는데 수정된 코드에서는 `csvOnlyCreateTarget.add()`가 사용되었습니다. 맞게 수정된 것인가요?

Kang-dot replied:
아마 실수한 것 같습니다. remove가 적용되도록 코드를 수정합니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -272,39 +279,69 @@ protected VerifyResultMessages checkCSVCfg(MigrationConfiguration config) {
+            List<String> exsistTableNameSet = existsTableNameList.get(schema);
+
+            boolean isOnlyCreateTargetExist = csvOnlyCreateTargetSet.retainAll(exsistTableNameSet);
+
+            if (isOnlyCreateTargetExist) {
                 throw new MigrationConfigurationCheckingErrorException(
-                        NLS.bind(Messages.objectMapPageErrMsgDuplicatedTable1, target));
+                        NLS.bind(
+                                Messages.objectMapPageErrMsgDuplicatedTable1,
+                                csvOnlyCreateTargetSet.get(0)));
```
`csvOnlyCreateTargetSet`과 `exsistTableNameSet`에서 중복된 값이 한 개도 없다면 retainAll()은 true를 반환하게 됩니다.
그러면서 원본 `csvOnlyCreateTargetSet`을 빈 set으로 변경하게 되는데, 이 상태에서 `csvOnlyCreateTargetSet.get(0))`을 호출하게 되면 IndexOutOfBoundsException 예외가 발생합니다.

버그 발생 가능성이 큰 코드라 생각되어 수정이 필요해 보입니다.

Kang-dot replied:
확인했습니다. 방어 코드를 적용했습니다

Srltas reviewed (commented):
마이그레이션 마법사 1단계에서 target을 Local CUBRID dump로 선택하면 5단계에서 에러가 발생합니다. target이 Local인 경우도 확인이 필요해 보입니다.

Kang-dot commented:
> 마이그레이션 마법사 1단계에서 target을 Local CUBRID dump로 선택하면 5단계에서 에러가 발생합니다. target이 Local인 경우도 확인이 필요해 보입니다.

팀 회의 결과 target이 local dump인 case는 다음 PR에서 수정하기로 결정되었습니다.

Kang-dot committed:
[TOOLS-4704]

- fix pr's review content

Kang-dot committed:
[TOOLS-4704] additional edit

- restore RmInvalidXMLCharReader.java file. it will edit in other issue
- remove empty line

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/view/SQLTableManageView.java:
```diff
@@ -377,7 +396,10 @@ protected void modify(
+                        } else if (columnIdx == 2) {
+                            super.modify(ti, element, columnIdx, value);
                         }
+
```
해당 줄은 의미상 줄바꿈이 꼭 필요해 보이지 않습니다.

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/page/view/SQLTableManageView.java:
```diff
@@ -588,6 +611,32 @@ public void run(IProgressMonitor monitor)
+        CellEditor[] cellEditorArray = tvSQL.getCellEditors();
+
+        ((ComboBoxCellEditor) cellEditorArray[2]).setItems(schemaNameArr);
+    }
```
수정이 안됐습니다.

Kang-dot replied:
누락 사항 수정 했습니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -272,39 +278,69 @@ protected VerifyResultMessages checkCSVCfg(MigrationConfiguration config) {
+            if (isReplaceTargetSet) {
+                sbWarn.append(
+                                Messages.bind(
+                                        Messages.msgWarnTableRecreated, csvReplaceTargetSet.get(0)))
```
csvReplaceTargetSet이 비어 있는 경우에 대한 예외 처리가 빠져있습니다.

Kang-dot replied:
확인했습니다. 누락된 코드를 추가했습니다

Srltas commented on com.cubrid.cubridmigration.ui/src/com/cubrid/cubridmigration/ui/wizard/utils/MigrationCfgUtils.java:
```diff
@@ -272,39 +278,69 @@ protected VerifyResultMessages checkCSVCfg(MigrationConfiguration config) {
+            List<String> exsistTableNameSet = existsTableNameList.get(schema);
+
+            boolean isOnlyCreateTargetExist = csvOnlyCreateTargetSet.retainAll(exsistTableNameSet);
+            if (isOnlyCreateTargetExist && !csvOnlyCreateTargetSet.isEmpty()) {
```
retainAll의 반환값을 통해 중복 여부를 판단하고 있는 부분은 의도와 다르게 동작할 가능성이 있어 보입니다.

`csvOnlyCreateTargetSet`과 `exsistTableNameSet`의 요소가 완전히 동일한 경우, **제거된 항목이 없기 때문에 false를 반환**하게 됩니다. 이 경우 실질적으로 모두 겹치는 값으로 판단해야 하지만, 현재 로직으로는 중복이 없다고 처리를 하게 됩니다.

이전 코드 흐름상 중복 값이 있을 경우 예외를 발생시키려는 의도가 있었던 걸로 보이기 때문에, 조건을 조금 더 명확하게 보완하거나 retainAll 사용을 한번 더 고려해볼 필요가 있습니다.

Kang-dot replied:
확인했습니다. retainAll 관련 코드를 stream의 anyMatch로 수정했습니다

Kang-dot committed:
[TOOLS-4704] change csv relate method

- change retain to anyMatch
- remove empty line

Kang-dot committed:
[TOOLS-4704] format code
